Rover/mode_dock.cpp

```diff
@@ -0,0 +1,252 @@
+        // accel_max of zero means no limit so use maximum acceleration
+        atc_accel_max = AR_DOCK_ACCEL_MAX;
+    }
+    const float accel_max = is_positive(g2.wp_nav.get_default_accel()) ? MIN(g2.wp_nav.get_default_accel(), atc_accel_max) : atc_accel_max;
+    const float jerk_max = is_positive(g2.wp_nav.get_default_jerk()) ? g2.wp_nav.get_default_jerk() : accel_max;
+
+    // initialise position controller
+    g2.pos_control.set_limits(speed_max, accel_max, g2.attitude_control.get_turn_lat_accel_max(), jerk_max);
+    g2.pos_control.init();
+
+    // set the position controller reversed in case the camera is mounted on vehicle's back
+    g2.pos_control.set_reversed(rover.precland.get_orient() == 4);
+
+    // construct unit vector in the desired direction from where we want the vehicle approach the dock
+    // this helps to dock the vehicle head-on even when we enter the dock mode at an angle towards the dock
+    _desired_heading_NE = Vector2f{cosf(radians(desired_dir)), sinf(radians(desired_dir))};
+
+    _loiter_after_docking = rover.is_boat();
+    _docking_complete = false;
+
+    return true;
+}
+
+void ModeDock::update()
+{
+    // if docking is complete, rovers stop and boats loiter
+    if (_docking_complete) {
+        // rovers stop, boats loiter 
+        // note that loiter update must be called after successfull initialisation on mode loiter
+        if (_loiter_after_docking) {
+            // mode loiter must be initialised before calling update_post_docking method
+            rover.mode_loiter.update();
+        } else {
+            stop_vehicle();
+        }
+        return;
+    }
+
+    bool real_dock_in_sight = rover.precland.get_target_position_cm(_dock_pos_rel_origin_cm);
+    Vector2f dock_pos_rel_vehicle_cm;
+    if (!calc_dock_pos_rel_vehicle_NE(dock_pos_rel_vehicle_cm)) {
+        g2.motors.set_throttle(0.0f);
+        g2.motors.set_steering(0.0f);
+        return;
+    }
+
+    _distance_to_destination = dock_pos_rel_vehicle_cm.length() * 0.01f;
+
+    // we force the vehicle to use real dock as target when we are too close to the dock
+    // note that heading correction does not work when we force real target
+    bool force_real_target = _distance_to_destination < _force_real_target_limit_cm * 0.01f;
+
+    // if we are close enough to the dock or the target is not in sight when we strictly require it
+    // we mark the docking to be completed so that the vehicle stops
+    if (_distance_to_destination <= stopping_dist || (force_real_target && !real_dock_in_sight)) {
+        _docking_complete = true;
+
+        // send a one time notification to GCS
+        gcs().send_text(MAV_SEVERITY_INFO, "Dock: Docking complete");
+
+        // initialise mode loiter if we need to loiter after docking completion
+        if (_loiter_after_docking) {
+            // if we fail to enter, we set _loiter_after_docking to false
+            _loiter_after_docking = rover.mode_loiter.enter();}
```

Comment:
I think the bracket should be on the next line.

Reply:
Oops! Yes, fixed it now